x-pack/plugins/security_solution/cypress/objects/rule.ts

```diff
@@ -46,20 +52,21 @@ export interface CustomRule {
   description: string;
   dataSource: RuleDataSource;
   interval?: string;
-  severity: string;
-  riskScore: string;
-  tags: string[];
+  severity?: string;
```

Comment:
Why the following properties are optional for custom query rules:

- `customQuery?: string;`
- `interval?: string;`
- `runsEvery?: Interval;`
- `lookBack?: Interval;`
- `severity?: string;`
- `riskScore?: string;`
- `tags?: string[];`

Reply:
These properties are part of the interface that represents the values of a rule to be filled on the UI:
* customQuery is optional because the user may fill the Custom query by importing it from a saved timeline
* And the other ones because are values you'll find filled by default on the page.

Reply:
@MadameSheema I think I understand the problem now.

Our e2e tests are driven by test data. Whatever is or is not in the data (e.g. an instance of `CustomRule`) will determine what actions will be taken in the UI. Example:

```ts
      visit(RULE_CREATION);
      fillDefineCustomRuleAndContinue(this.rule);
      fillAboutRuleAndContinue(this.rule);
      fillScheduleRuleAndContinue(this.rule);
```

```ts
export const fillDefineCustomRuleAndContinue = (rule: CustomRule | OverrideRule) => {
  if (rule.dataSource.type === 'dataView') {
    cy.get(DATA_VIEW_OPTION).click();
    cy.get(DATA_VIEW_COMBO_BOX).type(`${rule.dataSource.dataView}{enter}`);
  }
  fillCustomQuery(rule);
  cy.get(DEFINE_CONTINUE_BUTTON).should('exist').click({ force: true });
  cy.get(CUSTOM_QUERY_INPUT).should('not.exist');
};

export const fillScheduleRuleAndContinue = (rule: CustomRule | MachineLearningRule) => {
  if (rule.runsEvery) {
    cy.get(RUNS_EVERY_INTERVAL).type('{selectall}').type(rule.runsEvery.interval);
    cy.get(RUNS_EVERY_TIME_TYPE).select(rule.runsEvery.timeType);
  }
  if (rule.lookBack) {
    cy.get(LOOK_BACK_INTERVAL).type('{selectAll}').type(rule.lookBack.interval);
    cy.get(LOOK_BACK_TIME_TYPE).select(rule.lookBack.timeType);
  }
  cy.get(SCHEDULE_CONTINUE_BUTTON).click({ force: true });
};
```

This approach has a few downsides:

- We have conditional logic in functions like `fillScheduleRuleAndContinue`. This is commonly considered a bad practice in any automated tests because it increases complexity and non-determinism.
- It's a bit rigid: you can't customize the behavior for a particular test, because it is common to all tests and driven by data.
- The resulting actions that a test executes in the UI may not necessarily match the intention/expectation of the developer who writes this test. Or they may be redundant.

The suggestion is to make tests driven by code instead of data. By using single-purpose functions (Cypress tasks, commands, etc) and combining them into higher-order functions, tests will be implementing the exact logic each of them needs.

With that in place, test data can become just normal mock data where you normally have default values for properties you don't need.

Going back to the optional properties like `customQuery` and `runsEvery` - they all will be required and have some default values. Each test will decide:

- whether to use the default value and do something with it
- or override the default value and do something with it
- or just ignore it and do nothing

Reply:
I completely agree with you in all of this, since the tests were originally developed, complexity has increased over time making them:

* Less reliable
* Harder to maintain

I tried in the past to refactor them but I found hard stoppers from some colleagues. Extremely glad to collaborate on a refactor at least in your area. We can chat more about this :)